File: HW/HW12/HW12.java
import java.util.*;

public class HW12 {

    int[] parent;
    int[] rank;
    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        return

    }

    public int find(int x) {
        if (x != parent[x]) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    public boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX == rootY) {
            return false;
        }

        if (rank[rootX] > rank[rootY]) {
            parent[rootY] = rootX;
        } else if (rank[rootY] > rank[rootX]) {
            parent[rootX] = rootY;
        } else {
            parent[rootX] = rootY;
            rank[rootY]++;
        }
        return true;
    }
}

File: HW/HW12/HW12.py

def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
    parent = [i for i in range(n + 1)]
    rank = [0] * (n+1)

    def find(x):
        if x != parent[x]:
            parent[x]= find(parent[x])
        return parent[x]
    
    def union(x,y):
        rootX = find(x)
        rootY = find(y)

        if rootX == rootY:
            return False
        
        if rank[rootX] > rank[rootY]:
            parent[rootY] = rootX
        elif rank[rootY] > rank[rootX]:
            parent[rootX] = rootY
        else:
            parent[rootX] = rootY
            rank[rootY] += 1
        return True
